content/blog/2020-03-dplyr-1-0-0-new-verb-features/index.Rmarkdown

```diff
@@ -0,0 +1,172 @@
+
+```{r setup, message = FALSE}
+library(dplyr)
+packageVersion("dplyr") 
+```
+
+## Multiple rows and columns
+
+Two big changes make `summarise()` much more flexible. A single summary expression can now return:
+
+* A vector of any length, creating multiple rows.
+* A data frame, creating multiple columns.
+
+To get a sense for what this means, take this toy dataset:
+
+```{r}
+df <- tibble(
+  grp = rep(1:2, each = 5), 
+  x = c(rnorm(5, -0.25, 1), rnorm(5, 0, 1.5)),
+  y = c(rnorm(5, 0.25, 1), rnorm(5, 0, 0.5)),
+)
+df
+```
+
+You can now use summaries that return multiple values:
+
+```{r}
+df %>% 
+  group_by(grp) %>% 
+  summarise(rng = range(x))
+```
+
+Or return multiple columns from a single summary expression: 
+
+```{r}
+df %>% 
+  group_by(grp) %>% 
+  summarise(tibble(min = min(x), mean = mean(x)))
+```
+(This isn't very useful when used directly, but as you'll see shortly, it's really useful inside functions.)
+
+To put this another way, before dplyr 1.0.0, each summary had to be a single value (one row, one column), but now we've lifted that restriction so each summary can generate a rectangle of arbitrary size. This is a big change to `summarise()` but it should have minimal impact on existing code because it _broadens_ the interface: all existing code will continue to work, and a number of inputs that would have previously errored now work. 
+
+## Quantiles
+
+To demonstrate this new flexibility in a more useful situation, lets take a look at `quantile()`. `quantile()` was hard to use previously because it returns multiple values. Now it's straightforward:
+
+```{r}
+df %>% 
+  group_by(grp) %>% 
+  summarise(x = quantile(x, c(0.25, 0.5, 0.75)), q = c(0.25, 0.5, 0.75))
+```
+
+It would be nice to be able to reduce the duplication in this code so that we don't have to type the quantile values twice. We can now write a simple function because summary expressions can now be data frames or tibbles:
+
+```{r}
+quibble <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble(x = quantile(x, q), q = q)
+}
+df %>% 
+  group_by(grp) %>% 
+  summarise(quibble(x, c(0.25, 0.5, 0.75)))
+```
+
+In the past, one of the challenges of writing this sort of function was naming the columns. For example, when you call `quibble(y)` it'd be nice if you'd get columns `y` and `y_q`, not `x` and `x_q`. Now, thanks to the recent combination of [glue and tidy evaluation](https://www.tidyverse.org/blog/2020/02/glue-strings-and-tidy-eval/) that's easy to implement: 
+
+```{r}
+quibble2 <- function(x, q = c(0.25, 0.5, 0.75)) {
+  tibble("{{ x }}" := quantile(x, q), "{{ x }}_q" := q)
+}
+
+df %>% 
+  group_by(grp) %>% 
+  summarise(quibble2(y, c(0.25, 0.5, 0.75)))
+```
+
+One note of caution: name the output columns in a function like this is a surprisingly complex task, we're not yet sure what the best approach expect. Expect to here more about this as we continue to think about and experiment with it.
+
+## Data frame columns
```

Comment:
```suggestion
## Data-frame columns
```

I think you'll _stet_ me here, but my reasoning is as follows. Data frames have columns. They are the columns of data frames, but it's not unreasonable to reference a column of a data frame as a "data frame column." What is described below is columns that themselves contain/are of type data frame. They are data-frame columns, (which is why we call them **df-columns**, as you describe later).

Reply:
I bow your to hyphenatory wisdow

Reply:
😆